go/simple/v1/ampli/ampli.go

```diff
@@ -0,0 +1,385 @@
+// ampli.py
+//
+// Ampli - A strong typed wrapper for your Analytics
+//
+// This file is generated by Amplitude.
+// To update run 'ampli pull go-ampli'
+//
+// Required dependencies: analytics-go
+// Tracking Plan Version: 0
+// Build: 1.0.0
+// Runtime: go-ampli
+//
+// [View Tracking Plan]: https://data.amplitude.com/test-codegen/Test%20Codegen/events/main/latest
+//
+// [Full Setup Instructions]: https://data.amplitude.com/test-codegen/Test%20Codegen/implementation/main/latest/getting-started/go-ampli
+//
+
+package ampli
+
+import (
+	"log"
+
+	"github.com/amplitude/analytics-go/amplitude"
+)
+
+var Client = Ampli{}
+
+type Environment string
+
+const (
+	DEVELOPMENT Environment = "development"
+	PRODUCTION  Environment = "production"
+)
+
+var APIKey = map[Environment]string{
+	DEVELOPMENT: "",
+	PRODUCTION:  "",
+}
+
+var DefaultConfiguration = amplitude.Config{
+	FlushInterval:   amplitude.DefaultFlushInterval,
+	FlushQueueSize:  amplitude.DefaultFlushQueueSize,
+	FlushMaxRetries: amplitude.DefaultFlushMaxRetries,
+	Logger:          amplitude.NewDefaultLogger(),
+	MinIDLength:     amplitude.DefaultMinIDLength,
+	Callback:        nil,
+	ServerZone:      amplitude.ServerZoneUS,
+	UseBatch:        false,
+	Storage:         &amplitude.InMemoryStorage{},
+	OptOut:          false,
+	ServerURL:       amplitude.HTTPV2,
+	Plan: amplitude.Plan{
+		Branch:    "main",
+		Source:    "python-Ampli",
+		Version:   "0",
+		VersionID: "79154a50-f057-4db5-9755-775e4e9f05e6",
+	},
+}
+
+// LoadClientOptions is Client options setting to initialize Ampli client.
+//
+// Params:
+// 	- APIKey: the API key of Amplitude project
+// 	- Instance: the core SDK instance used by Ampli client
+// 	- Configuration: the core SDK client configuration instance
+type LoadClientOptions struct {
+	APIKey        string
+	Instance      amplitude.Client
+	Configuration amplitude.Config
+}
+
+func (l LoadClientOptions) isEmpty() bool {
+	return l.APIKey == "" && l.Instance == nil && l.Configuration.IsEmpty()
+}
+
+// LoadOptions is options setting to initialize Ampli client.
+//
+// Params:
+//	- Environment: the environment of Amplitude Data project
+//	- Disabled: the flag of disabled Ampli client
+//	- Client: the LoadClientOptions struct
+type LoadOptions struct {
+	Environment Environment
+	Disabled    bool
+	Client      LoadClientOptions
+}
+
+// Identify
+//
+// [View in Tracking Plan]: https://data.amplitude.com/test-codegen/Test%20Codegen/events/main/latest/Identify
+//
+// Identify properties
+//
+// Params:
+//	- RequiredNumber: description for Ampli.Identify RequiredNumber
+//	- OptionalArray: description for Ampli.Identify OptionalArray
+type Identify struct {
```

Comment:
In case of `Go` required and optional fields can't be distinguished. I think, event builders (like in Java) should be used instead

Reply:
Added `NewStronglyTypedEvent()` function for every strongly typed event, which takes required properties as arguments

Reply:
For example, if we have an event with optional `optionalBoolean` property:
in JS/TS we have 3 options:
1. `event = {}` - `optionalBoolean` is undefined and isn't included in payload
2. `event = { optionalBooelan: false }` - `optionalBoolean` is `false` and included in payload
3. `event = { optionalBooelan: true }` - `optionalBoolean` is `true` and included in payload

With Go struct we do not have the option 1:
1. `Event{}` - `optionalBoolean` is `false` and included in payload
2. `Event{ OptionalBoolean: false }` - `optionalBoolean` is `false` and included in payload
3. `Event{ OptionalBoolean: true }` - `optionalBoolean` is `true` and included in payload

So we need something like Java builders to allow the option 1 to not include `optionalBoolean` to payload.

Reply:
Added a setter for every optional properties, for example `SetOptionalBoolean(optionalBoolean bool)`. Also added a private `optionals map[string]bool` property in struct definition to keep track of which optional properties are set. It is used in `ToEvent()` to decide whether to add the optional property to `EventProperties` or not. 

Wrote some tests in `ampli_test.go` to include the options in your comments.

Reply:
1. maybe Builder pattern is more preferred to be consistent with Java codegen
2. possible alternative implementation:
```
type baseEvent struct {
	eventType  string
	properties map[string]interface{}
}

type amplitudeEvent interface {
	toAmplitudeEvent() amplitude.Event
}

func newBaseEvent(eventType string, properties map[string]interface{}) *baseEvent {
	return &baseEvent{
		eventType:  eventType,
		properties: properties,
	}
}

func (event *baseEvent) toAmplitudeEvent() amplitude.Event {
	return amplitude.Event{
		EventType:       event.eventType,
		EventProperties: event.properties,
	}
}

type Identify struct {
	*baseEvent
}

func NewIdentify(requiredNumber float64) *Identify {
	return &Identify{
		newBaseEvent(amplitude.IdentifyEventType, map[string]interface{}{
			"requiredNumber": requiredNumber,
		}),
	}
}

func (event *Identify) SetOptionalArray(optionalArray []string) *Identify {
	event.properties["optionalArray"] = optionalArray
	return event
}

func (event *Identify) toAmplitudeEvent() amplitude.Event {
	identify := amplitude.Identify{}
	for name, value := range event.properties {
		identify.Set(name, value)
	}

	return amplitude.Event{
		EventType:      event.eventType,
		UserProperties: identify.Properties,
	}
}

type Group struct {
	*baseEvent
	groups map[string][]string
}

func NewGroup(requiredBoolean bool) *Group {
	return &Group{
		baseEvent: newBaseEvent(amplitude.GroupIdentifyEventType, map[string]interface{}{
			"requiredBoolean": requiredBoolean,
		}),
	}
}

func (event *Group) SetOptionalString(optionalString string) *Group {
	event.properties["optionalString"] = optionalString
	return event
}

func (event *Group) toAmplitudeEvent() amplitude.Event {
	identify := amplitude.Identify{}
	for name, value := range event.properties {
		identify.Set(name, value)
	}

	return amplitude.Event{
		EventType:       event.eventType,
		GroupProperties: identify.Properties,
		Groups:          event.groups,
	}
}

type EventWithAllProperties struct {
	*baseEvent
}

type RequiredEnum string

const (
	RequiredEnumEnum1 RequiredEnum = "Enum1"
	RequiredEnumEnum2 RequiredEnum = "Enum2"
)

func NewEventWithAllProperties(requiredArray []string, requiredBool bool, requiredEnum RequiredEnum, requiredInteger int, requiredNumber float64, requiredString string) *EventWithAllProperties {
	return &EventWithAllProperties{
		newBaseEvent(amplitude.GroupIdentifyEventType, map[string]interface{}{
			"requiredArray":   requiredArray,
			"requiredBool":    requiredBool,
			"requiredEnum":    requiredEnum,
			"requiredInteger": requiredInteger,
			"requiredNumber":  requiredNumber,
			"requiredString":  requiredString,
		}),
	}
}

func (event *EventWithAllProperties) SetOptionalString(optionalString string) *EventWithAllProperties {
	event.properties["optionalString"] = optionalString
	return event
}

func (event *EventWithAllProperties) SetOptionalBool(optionalBool bool) *EventWithAllProperties {
	event.properties["optionalBool"] = optionalBool
	return event
}
```

Reply:
That's awesome! 
- More consistent with Java builder pattern 
- No `optionals` in strongly typed events
- Use only one `toAmplitudeEvent()` 

Have a small question here. Why do we use pointer here
```
type EventWithAllProperties struct {
	*baseEvent
}
```
instead of 
```
type EventWithAllProperties struct {
	baseEvent
}
```
I've tried and both of them work here.

Updated

Reply:
Sorry, it was alternative implementation for your current code, not the `Builder` pattern.
The `Builder` pattern is implemented in Java code:
https://github.com/amplitude/ampli-examples/blob/main/jre/java/AmpliApp/src/main/java/com/amplitude/ampli/EventWithAllProperties.java
https://github.com/amplitude/ampli-examples/blob/main/jre/java/AmpliApp/src/main/java/org/example/App.java#L107

Reply:
`baseEvent`/`*baseEvent` doesn't matter in this case, you're right.

Reply:
I actually quite like this implementation! Specially it uses only one `toAmplitudeEvent()`, so less code in codegen